x-pack/filebeat/input/mito/config.go

```diff
@@ -0,0 +1,178 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package mito
+
+import (
+	"errors"
+	"fmt"
+	"net/http"
+	"net/url"
+	"regexp"
+	"time"
+
+	"github.com/elastic/beats/v7/libbeat/common/transport/httpcommon"
+)
+
+// config is the top-level configuration for a mito input.
+type config struct {
+	// Interval is the period interval between runs of the input.
+	Interval time.Duration `config:"interval" validate:"required"`
+
+	// Program is the CEL program to be run for each polling.
+	Program string `config:"program" validate:"required"`
+	// Regexps is the set of regular expression to be made
+	// available to the program.
+	Regexps map[string]string `config:"regexp"`
+	// State is the initial state to be provided to the
+	// program. If it has a cursor field, that field will
+	// be overwritten by any stored cursor, but will be
+	// available if no stored cursor exists.
+	State map[string]interface{} `config:"state"`
+
+	// Auth is the authentication config for connection to an HTTP
+	// API endpoint.
+	Auth authConfig `config:"auth"`
+
+	// Resource is the configuration for establishing an
+	// HTTP request or for locating a local resource.
+	Resource *ResourceConfig `config:"resource" validate:"required"`
+}
+
+func (c config) Validate() error {
+	if c.Interval <= 0 {
+		return errors.New("interval must be greater than 0")
+	}
+	_, err := regexpsFromConfig(c)
+	if err != nil {
+		return fmt.Errorf("failed to check regular expressions: %w", err)
+	}
+	// TODO: Consider just building the program here to avoid this wasted work.
```

Comment:
You could add a type that implements [Unpack](https://pkg.go.dev/github.com/elastic/go-ucfg#StringUnpacker) to do the work of validating and building the program.

Reply:
The issue that I was struggling with with this is that this then depends on having all the auth work also validated in the config validation for the program since the functional program needs the http client to be ready when the program is compiled. The cost of compiling and checking the program is not huge any only happens once per run of the input, so I fell on the side of keeping the code simpler and more modular at the slight additional cost of compiling twice. If I can see a nice way of threading this camel in, I'll certainly revisit.